File: spikes/ts/src/sched.ts
// SP-5: tempo, casualità e ordinamento sono iniettabili (V29).
// Nessun Date.now(), nessun Math.random() (C5).

export class Rng {
  private s: number;
  constructor(seed: number) {
    const s = (seed * 1664525 + 1013904223) >>> 0;
    // xorshift resta bloccato su zero: la guardia evita un seed morto.
    this.s = s === 0 ? 1 : s;
  }
  next(): number {
    let x = this.s;
    x ^= x << 13;
    x >>>= 0;
    x ^= x >>> 7;
    x ^= x << 17;
    x >>>= 0;
    this.s = x;
    return x;
  }
  below(n: number): number {
    return this.next() % n;
  }
}

interface Task {
  nome: string;
  rimanenti: number;
}

export class World {
  private rng: Rng;
  private orologio = 0;
  private tasks: Task[] = [];
  private traccia: string[] = [];

  constructor(seed: number) {
    this.rng = new Rng(seed);
  }

  spawn(nome: string, steps: number): void {
    this.tasks.push({ nome, rimanenti: steps });
  }

  /** Avanza l'orologio virtuale senza attendere davvero (C3). */
  sleepVirtuale(ms: number): void {
    this.orologio += ms;
  }

  now(): number {
    return this.orologio;
  }

  run(): string[] {
    while (this.tasks.length > 0) {
      const i = this.rng.below(this.tasks.length);
      this.orologio += 1;

      const guasto = this.rng.below(20) === 0;
      const nome = this.tasks[i]!.nome;

      if (guasto) {
        this.traccia.push(`t=${this.orologio} ${nome} GUASTO`);
        this.tasks.splice(i, 1);
        continue;
      }

      this.traccia.push(`t=${this.orologio} ${nome} passo`);
      this.tasks[i]!.rimanenti -= 1;
      if (this.tasks[i]!.rimanenti === 0) {
        this.traccia.push(`t=${this.orologio} ${nome} fine`);
        this.tasks.splice(i, 1);
      }
    }
    return this.traccia;
  }
}

File: spikes/rust/src/sched.rs
//! SP-5: tempo, casualità e ordinamento delle attività sono iniettabili (V29).
//! Nessuna lettura dell'orologio di sistema, nessun RNG globale (C5).

/// Generatore deterministico. Sostituisce ogni fonte di casualità.
pub(crate) struct Rng(u64);

impl Rng {
    pub(crate) fn new(seed: u64) -> Self {
        let s = seed.wrapping_mul(6364136223846793005).wrapping_add(1);
        // xorshift resta bloccato su zero: la guardia evita un seed morto.
        Rng(if s == 0 { 1 } else { s })
    }
    fn next(&mut self) -> u64 {
        // xorshift64: deterministico e sufficiente per uno spike
        let mut x = self.0;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.0 = x;
        x
    }
    pub(crate) fn below(&mut self, n: u64) -> u64 {
        self.next() % n
    }
}

struct Task {
    nome: String,
    rimanenti: u32,
}

/// Il "mondo" possiede tempo, casualità e coda delle attività.
pub struct World {
    rng: Rng,
    orologio: u64,
    tasks: Vec<Task>,
    traccia: Vec<String>,
}

impl World {
    pub fn new(seed: u64) -> Self {
        World {
            rng: Rng::new(seed),
            orologio: 0,
            tasks: Vec::new(),
            traccia: Vec::new(),
        }
    }

    pub fn spawn(&mut self, nome: &str, steps: u32) {
        self.tasks.push(Task {
            nome: nome.to_string(),
            rimanenti: steps,
        });
    }

    /// Avanza l'orologio virtuale senza attendere davvero (C3).
    pub fn sleep_virtuale(&mut self, millis: u64) {
        self.orologio += millis;
    }

    pub fn now(&self) -> u64 {
        self.orologio
    }

    /// Esegue fino a esaurimento, interlacciando le attività secondo il seed.
    pub fn run(&mut self) -> Vec<String> {
        while !self.tasks.is_empty() {
            let i = self.rng.below(self.tasks.len() as u64) as usize;
            self.orologio += 1;

            // Guasto iniettato in un punto scelto dal seed (C4).
            let guasto = self.rng.below(20) == 0;
            let nome = self.tasks[i].nome.clone();

            if guasto {
                self.traccia
                    .push(format!("t={} {} GUASTO", self.orologio, nome));
                self.tasks.remove(i);
                continue;
            }

            self.traccia
                .push(format!("t={} {} passo", self.orologio, nome));
            self.tasks[i].rimanenti -= 1;
            if self.tasks[i].rimanenti == 0 {
                self.traccia
                    .push(format!("t={} {} fine", self.orologio, nome));
                self.tasks.remove(i);
            }
        }
        self.traccia.clone()
    }
}
